src/eval/mod.rs

```diff
@@ -154,11 +154,12 @@ impl<R: ImportResolver> VirtualMachine<R> {
         }
     }
 
-    /// Reset the state of the machine (stacks and eval mode) to prepare for another evaluation round.
+    /// Reset the state of the machine (stacks, eval mode and thunk state) to prepare for another evaluation round.
     pub fn reset(&mut self) {
         self.eval_mode = Default::default();
         self.call_stack.0.clear();
-        self.stack.clear();
+        self.stack.reset_thunks();
+        self.stack.clear(); //Do we even need to call this at this point?
```

Comment:
Well, you either clear inside `reset_thunks` (but then maybe the name is not really adapted anymore, because it's doing more?) as you do now, and you don't have to call to `clear()`. Or you make `reset_thunks` more atomic, by just iterating immutably on the stack, and keep the separate `clear()` function. I don't have a strong opinion, beside the naming issue.